lib/libexfat.c

```diff
@@ -358,26 +358,27 @@ off_t exfat_get_root_entry_offset(struct exfat_blk_dev *bd)
 {
 	struct pbr *bs;
 	int nbytes;
-	unsigned int cluster_size;
+	unsigned int cluster_size, sector_size;
 	off_t root_clu_off;
 
-	bs = (struct pbr *)malloc(sizeof(struct pbr));
+	bs = (struct pbr *)malloc(EXFAT_MAX_SECTOR_SIZE);
 	if (!bs) {
 		exfat_err("failed to allocate memory\n");
 		return -ENOMEM;
 	}
 
-	nbytes = exfat_read(bd->dev_fd, bs, sizeof(struct pbr), 0);
-	if (nbytes != sizeof(struct pbr)) {
+	nbytes = exfat_read(bd->dev_fd, bs, EXFAT_MAX_SECTOR_SIZE, 0);
+	if (nbytes != EXFAT_MAX_SECTOR_SIZE) {
 		exfat_err("boot sector read failed: %d\n", errno);
 		free(bs);
 		return -1;
 	}
 
-	cluster_size = (1 << bs->bsx.sect_per_clus_bits) * bd->sector_size;
-	root_clu_off = le32_to_cpu(bs->bsx.clu_offset) * bd->sector_size +
-		le32_to_cpu(bs->bsx.root_cluster - EXFAT_RESERVED_CLUSTERS)
-		* cluster_size;
+	sector_size = 1 << bs->bsx.sect_size_bits;
+	cluster_size = (1 << bs->bsx.sect_per_clus_bits) * sector_size;
+	root_clu_off = le32_to_cpu(bs->bsx.clu_offset) * sector_size +
+		le32_to_cpu(bs->bsx.root_cluster - EXFAT_RESERVED_CLUSTERS) *
```

Comment:
I think we have to change this le32_to_cpu call to le32_to_cpu(bs->bsx.root_cluster) - EXFAT_RESERVED_CLUSTERS.

Reply:
Yes, but it is not related with this patch subject. You can fix it with another patch.